ast/compile.go

```diff
@@ -883,6 +883,110 @@ func compileSchema(goSchema interface{}) (*gojsonschema.Schema, error) {
+		if result.Types.String() != schemas[i].Types.String() {
+			return nil, fmt.Errorf("unable to merge these schemas")
+		}
+		if result.Types.Contains("object") && len(result.PropertiesChildren) > 0 && schemas[i].Types.Contains("object") && len(schemas[i].PropertiesChildren) > 0 {
+			for j := 0; j < len(schemas[i].PropertiesChildren); j++ {
+				result.PropertiesChildren = append(result.PropertiesChildren, schemas[i].PropertiesChildren[j])
+			}
+		} else if result.Types.Contains("array") && len(result.ItemsChildren) > 0 && schemas[i].Types.Contains("array") && len(schemas[i].ItemsChildren) > 0 {
+			for j := 0; j < len(schemas[i].ItemsChildren); j++ {
+				if len(result.ItemsChildren)-1 < j && !(len(schemas[i].ItemsChildren)-1 < j) {
+					result.ItemsChildren = append(result.ItemsChildren, schemas[i].ItemsChildren[j])
+				}
+				if result.ItemsChildren[j].Types.String() != schemas[i].ItemsChildren[j].Types.String() {
+					return nil, fmt.Errorf("unable to merge these schemas due to different types")
+				} else if result.ItemsChildren[j].Types.String() == schemas[i].ItemsChildren[j].Types.String() {
+					result.ItemsChildren[j].Types = schemas[i].ItemsChildren[j].Types
+				}
+			}
+		} else if result.Types.String() != schemas[i].Types.String() {
+			return nil, fmt.Errorf("unable to merge these schemas")
+		} else if result.Types.Contains("string") && schemas[i].Types.Contains("string") {
+			stringPresent := true
+			for k := i + 1; k < len(schemas); k++ {
+				if schemas[k].Types.Contains("string") {
+					stringPresent = true
+				} else {
+					stringPresent = false
+				}
+			}
+			if stringPresent {
+				return result, nil
+			} else {
+				return nil, fmt.Errorf("unable to merge these schemas")
+			}
+		} else if result.Types.Contains("integer") && schemas[i].Types.Contains("integer") {
+			integerPresent := true
+			for k := i + 1; k < len(schemas); k++ {
+				if schemas[k].Types.Contains("integer") {
+					integerPresent = true
+				} else {
+					integerPresent = false
+				}
+			}
+			if integerPresent {
+				return result, nil
+			} else {
+				return nil, fmt.Errorf("unable to merge these schemas")
+			}
+		} else if result.Types.Contains("boolean") && schemas[i].Types.Contains("boolean") {
+			booleanPresent := true
+			for k := i + 1; k < len(schemas); k++ {
+				if schemas[k].Types.Contains("boolean") {
+					booleanPresent = true
+				} else {
+					booleanPresent = false
+				}
+			}
+			if booleanPresent {
+				return result, nil
+			} else {
+				return nil, fmt.Errorf("unable to merge these schemas")
+			}
+		} else if result.Types.Contains("number") && schemas[i].Types.Contains("number") {
+			numberPresent := true
+			for k := i + 1; k < len(schemas); k++ {
+				if schemas[k].Types.Contains("number") {
+					numberPresent = true
+				} else {
+					numberPresent = false
+				}
+			}
+			if numberPresent {
+				return result, nil
+			} else {
+				return nil, fmt.Errorf("unable to merge these schemas")
+			}
+		}
```

Comment:
I dont think you need any of this code. 
This is my reasoning:
If result has type string, any other type in schemas[i] will be rejected by the first if check and marked as unmergeable. If all are strings in schemas[i]. then result is still type string and thats all you need.


So you will need to remove the if else below on https://github.com/aavarghese/opa/pull/1/files?file-filters%5B%5D=.go#diff-067a842cc5e1a348551f4c7656120105acedafb2faac6e7b7ff57b75c5349570R983-R987
and just have a `return result, nil` --> you should return result for all result types not just object or array.